telemetry/jetbrains/src/main/kotlin/software/aws/toolkits/telemetry/generator/GenerationFunctions.kt

```diff
@@ -0,0 +1,164 @@
+    addComment("THIS FILE IS GENERATED! DO NOT EDIT BY HAND!")
+}
+
+fun FileSpec.Builder.generateTelemetryEnumTypes(items: List<TelemetryMetricType>) {
+    items.forEach {
+        // We only need to generate enums if they are actually enums, skip other types
+        if (it.allowedValues == null) {
+            return@forEach
+        }
+        generateTelemetryEnumType(it)
+    }
+}
+
+fun FileSpec.Builder.generateTelemetryEnumType(item: TelemetryMetricType) {
+    val enum = TypeSpec.enumBuilder(item.name.toTypeFormat())
+        .primaryConstructor(
+            FunSpec.constructorBuilder()
+                .addParameter("value", String::class)
+                .build()
+        )
+        .addProperty(PropertySpec.builder("value", String::class).initializer("value").build())
+        .addFunction(FunSpec.builder("toString").addModifiers(KModifier.OVERRIDE).returns(String::class).addStatement("return value").build())
+        .addKdoc(item.description)
+
+    item.allowedValues!!.forEach { enumValue ->
+        enum.addEnumConstant(
+            enumValue.toString().toUpperCase().filterInvalidCharacters(), TypeSpec.anonymousClassBuilder()
+                .addSuperclassConstructorParameter("%S", enumValue.toString())
+                .build()
+        )
+    }
+
+    val companion = TypeSpec.companionObjectBuilder()
+        .addFunction(
+            FunSpec.builder("from")
+                .returns(ClassName("", item.name.toTypeFormat()))
+                .addParameter("type", Any::class)
+                .addStatement("return values().first { it.value == type.toString() }")
+                .build()
+        )
+        .build()
+
+    enum.addType(companion)
+
+    addType(enum.build())
+}
+
+fun FileSpec.Builder.generateTelemetryObjects(item: TelemetryDefinition) =
+    item
+        .metrics
+        .sortedBy { it.name }
+        .groupBy { it.name.split("_").first().toLowerCase() }
+        .forEach { metrics: Map.Entry<String, List<Metric>> -> generateNamespaces(item.types!!, metrics.key, metrics.value) }
+
+fun FileSpec.Builder.generateNamespaces(types: List<TelemetryMetricType>, namespaceType: String, metrics: List<Metric>) {
+    val namespace = TypeSpec.objectBuilder("${namespaceType.toTypeFormat()}Telemetry")
+    metrics.forEach { namespace.generateRecordFunctions(it, types) }
+    addType(namespace.build())
+}
+
+fun TypeSpec.Builder.generateRecordFunctions(metric: Metric, types: List<TelemetryMetricType>) {
+    // metric.name.split("_")[1] is guaranteed to work at this point because the schema requires the metric name to have at least 1 underscore
+    val functionName = metric.name.split("_")[1]
+    val parameters = buildParameters(metric, types)
+    val functionBuilder = FunSpec.builder(functionName)
+    functionBuilder.addParameters(parameters)
+    functionBuilder.generateFunctionBody(metric)
+    functionBuilder.addKdoc(metric.description)
+    val function = functionBuilder.build()
+    addFunction(function)
```

Comment:
inline above line, add white space before we start the overload logic

Reply:
function is used below but added line return